crates/composefs/src/fsverity/mod.rs

```diff
@@ -72,10 +113,116 @@ pub fn compute_verity<H: FsVerityHashValue>(data: &[u8]) -> H {
+/// attempts.
+pub fn enable_verity_with_retry<H: FsVerityHashValue>(
+    fd: impl AsFd,
+) -> Result<(), EnableVerityError> {
+    let mut attempt = 1;
+    loop {
+        match enable_verity_raw::<H>(&fd) {
+            Err(EnableVerityError::FileOpenedForWrite) if attempt < 3 => {
+                std::thread::sleep(std::time::Duration::from_millis(1));
+                attempt += 1;
+            }
+            other => return other,
+        }
+    }
+}
+
+/// Enable fs-verity on the given file.  If the given file cannot be
+/// enabled because it is opened as writable, then a new copy of the
+/// file will be returned instead.  No attempt is made to sync the
+/// copied file contents to disk, it is up to the caller to do so if
+/// desired.
+///
+/// Take special note that in the case where a copied file descriptor
+/// is returned, the returned file is created as a tempfile and is
+/// unlinked.  Presumably the caller should take care to make this
+/// file permanent, using a combination of `linkat` and `renameat` to
+/// replace the original file.
+///
+/// This uses `enable_verity_raw()` and `enable_verity_with_retry()`
+/// and is subject to the same restrictions.
+///
+/// # Arguments:
+/// * `dirfd`: A directory file descriptor, used to determine the placement (via O_TMPFILE) of the new file (if necessary).
+/// * `fd`: The file decriptor to enable verity on
+pub fn enable_verity_maybe_copy<H: FsVerityHashValue>(
+    dirfd: impl AsFd,
+    fd: OwnedFd,
+    mode: Mode,
+) -> Result<VerityFd, EnableVerityError> {
+    match enable_verity_with_retry::<H>(&fd) {
+        Ok(_) => Ok(VerityFd::Orig(fd)),
+        Err(EnableVerityError::FileOpenedForWrite) => {
+            enable_verity_on_copy::<H>(dirfd, fd, mode).map(VerityFd::Copy)
+        }
+        Err(other) => Err(other),
+    }
+}
+
+/// Enable fs-verity on a new copy of `fd`, consuming `fd` and
+/// returning the new copy.  The copy is created via O_TMPFILE
+/// relative to `dirfd`.
+fn enable_verity_on_copy<H: FsVerityHashValue>(
+    dirfd: impl AsFd,
+    fd: OwnedFd,
+    mode: Mode,
+) -> Result<OwnedFd, EnableVerityError> {
+    let mut fd = File::from(fd);
+
+    loop {
+        fd.rewind().map_err(EnableVerityError::Rewind)?;
+
+        let mut new_rw_fd = File::from(
+            openat(
+                &dirfd,
+                ".",
+                OFlags::CLOEXEC | OFlags::RDWR | OFlags::TMPFILE,
+                mode,
+            )
+            .map_err(|e| EnableVerityError::Openat(e.into()))?,
+        );
+
+        std::io::copy(&mut fd, &mut new_rw_fd)?;
+        let new_ro_fd = open(
+            proc_self_fd(&new_rw_fd),
+            OFlags::RDONLY | OFlags::CLOEXEC,
+            Mode::empty(),
+        )
+        .map_err(|e| EnableVerityError::Open(e.into()))?;
+        drop(new_rw_fd);
+        if enable_verity_with_retry::<H>(&new_ro_fd).is_ok() {
+            return Ok(new_ro_fd);
+        }
```

Comment:
```suggestion
        }
        fd = new_ro_fd;
```

I think we need that to be robust against long running forks.

Reply:
At this point we've already determined that we have to make a copy (`enable_verity_on_copy` is internal, called only when we have exhausted all other efforts and we are pretty sure there *is* a long running fork, otherwise the `enable_verity_with_retry` variant would have caught it).  The original `fd` doesn't matter other than providing us a source to copy the data from.  In other words, it doesn't matter if a long running fork has a rw copy of it, we're not trying to enable verity on that file.